src/System.Private.CoreLib/shared/System/Collections/HashHelpers.cs

```diff
@@ -86,5 +88,30 @@ public static int ExpandPrime(int oldSize)
 
             return GetPrime(newSize);
         }
+
+#if BIT64
+        public static ulong GetFastModMultiplier(uint divisor)
+            => ulong.MaxValue / divisor + 1;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe uint FastMod(uint value, uint divisor, ulong multiplier)
+        {
+            // 64bit * 64bit => 128bit isn't currently supported by Math https://github.com/dotnet/corefx/issues/41822
+            if (Bmi2.X64.IsSupported)
+            {
+                // Using fastmod from Daniel Lemire https://lemire.me/blog/2019/02/08/faster-remainders-when-the-divisor-is-a-constant-beating-compilers-and-libdivide/
+
+                ulong lowbits = multiplier * value;
+                ulong low;
+                ulong high = Bmi2.X64.MultiplyNoFlags(lowbits, divisor, &low);
+                return (uint)high;
+            }
+            else
+            {
+                // 64bit * 64bit => 128bit isn't supported so we will use the slower modulo.
+                return value % divisor;
```

Comment:
What would be performance if you use two '64-bit * 64-bit => 64-bit' multiplications instead of the modulo?
``` C#
return (uint)((((ulong)(uint)lowbits * divisor >> 32) + (lowbits >> 32) * divisor) >> 32);
```

Reply:
Is this valid? If so I'd probably prefer this to using `Bmi2.X64.MultiplyNoFlags` and revisit when `Math.MultiplyHigh` is available as it gets around the Bing concern about the Support check being baked in and not being reJitted to have it removed https://github.com/dotnet/coreclr/pull/27149#discussion_r334462499

> Dictionary is one of the hottest pieces of code we run and we run with tiered jitting disabled

Reply:
> Is this valid?

Sure. Considering 32 bits as one 'digit', we are multiplying two-digits `lowbits` and one-digit `divisor` using the school method. We need only the third digit (hundreds place) of the product. But you might add a `Debug.Assert` just in case. :)

Reply:
Running perf tests
```diff
-    mov      r14d, dword ptr [rbx+8]
-    mov      eax, ebp
-    xor      rdx, rdx
-    div      edx:eax, r14d
-    cmp      edx, r14d
-    jae      G_M10972_IG15
-    movsxd   rax, edx
-    mov      ebx, dword ptr [rbx+4*rax+16]
-    mov      r14, gword ptr [rdi+16]
+    mov      rax, gword ptr [rdi+8]
+    mov      edx, dword ptr [rax+8]
+    mov      ecx, edx
+    mov      r8, qword ptr [rdi+48]
+    mov      r9d, ebx
+    imul     r8, r9
+    mov      r9d, r8d
+    mov      r10d, ecx
+    imul     r9, r10
+    shr      r9, 32
+    shr      r8, 32
+    mov      ecx, ecx
+    imul     rcx, r8
+    add      rcx, r9
+    shr      rcx, 32
+    cmp      ecx, edx
+    jae      G_M10976_IG15
+    movsxd   rdx, ecx
+    lea      rax, bword ptr [rax+4*rdx+16]
+    mov      ebp, dword ptr [rax]
+    mov      rdi, gword ptr [rdi+16]
```

Reply:
Still a improvement, though lesser
```
|                       Method |   Toolchain |      Mean | Ratio |
|----------------------------- |------------ |----------:|------:|
|    TryGetValue_17_Int_Object |        base |  8.427 ns |  1.00 |
|    TryGetValue_17_Int_Object |        diff |  7.070 ns |  0.84 |
|    TryGetValue_3k_Int_Object |        base |  7.611 ns |  1.00 |
|    TryGetValue_3k_Int_Object |        diff |  6.635 ns |  0.87 |
|    TryGetValue_1M_Int_Object |        base |  7.743 ns |  1.00 |
|    TryGetValue_1M_Int_Object |        diff |  6.803 ns |  0.88 |
|                              |             |           |       |
| TryGetValue_17_String_Object |        base | 16.233 ns |  1.00 |
| TryGetValue_17_String_Object |        diff | 14.775 ns |  0.91 |
| TryGetValue_3k_String_Object |        base | 22.777 ns |  1.00 |
| TryGetValue_3k_String_Object |        diff | 21.387 ns |  0.94 |
| TryGetValue_1M_String_Object |        base | 74.274 ns |  1.00 |
| TryGetValue_1M_String_Object |        diff | 72.219 ns |  0.97 |
|                              |             |           |       |
| TryGetValue_17_Object_Object |        base | 17.341 ns |  1.00 |
| TryGetValue_17_Object_Object |        diff | 17.396 ns |  1.01 |
| TryGetValue_3k_Object_Object |        base | 24.407 ns |  1.00 |
| TryGetValue_3k_Object_Object |        diff | 22.979 ns |  0.94 |
| TryGetValue_1M_Object_Object |        base | 96.327 ns |  1.00 |
| TryGetValue_1M_Object_Object |        diff | 88.482 ns |  0.92 |
```

Reply:
KNucleotide_1 is ok
```
|        Method | Toolchain |     Mean | Ratio |
|-------------- |---------- |---------:|------:|
| KNucleotide_1 |      base | 378.8 ms |  1.00 |
| KNucleotide_1 |      diff | 352.4 ms |  0.93 |
```
KNucleotide_9 is a bit odder
```
|        Method | Toolchain |     Mean |   Median |      Min |       Max | Ratio |
|-------------- |---------- |---------:|---------:|---------:|----------:|------:|
| KNucleotide_9 |      base | 92.18 ms | 92.64 ms | 85.09 ms | 100.73 ms |  1.00 |
| KNucleotide_9 |      diff | 92.18 ms | 92.49 ms | 86.86 ms |  95.95 ms |  1.00 |
```

Reply:
@benaadams Thank you for measuring. Sure, the intrinsic would be faster. I considered this only as a fallback.

Reply:
Have gone for this. The downside with the intrinsic is the `.IsSupported` stub and branch and that its not x-arch. Can change it if/when `Math.MultiplyHigh` appears.

Reply:
Have gone for this. The downside with the intrinsic is the `.IsSupported` stub and branch and that its not x-arch. Can change it if/when `Math.MultiplyHigh` appears. (As standard multiply is already 128bit just the hit bits aren't exposed in .NET currently)